flow/connectors/utils/aws.go

```diff
@@ -20,79 +24,158 @@ type AWSSecrets struct {
-	AccessKeyID     string `json:"accessKeyId"`
-	SecretAccessKey string `json:"secretAccessKey"`
-	Region          string `json:"region"`
-	BucketPath      string `json:"bucketPath"`
+type ClickHouseS3Credentials struct {
+	Provider   AWSCredentialsProvider
+	BucketPath string
 }
 
-func GetAWSSecrets(creds S3PeerCredentials) (*AWSSecrets, error) {
-	awsRegion := creds.Region
-	if awsRegion == "" {
-		awsRegion = os.Getenv("AWS_REGION")
-	}
-	if awsRegion == "" {
-		return nil, errors.New("AWS_REGION must be set")
-	}
+type AWSCredentials struct {
+	AWS         aws.Credentials
+	EndpointUrl *string
+}
 
-	awsEndpoint := creds.Endpoint
-	if awsEndpoint == "" {
-		awsEndpoint = os.Getenv("AWS_ENDPOINT")
-	}
+type AWSCredentialsProvider interface {
+	Retrieve(ctx context.Context) (AWSCredentials, error)
+	GetUnderlyingProvider() aws.CredentialsProvider
+	GetRegion() string
+}
 
-	awsKey := creds.AccessKeyID
-	if awsKey == "" {
-		awsKey = os.Getenv("AWS_ACCESS_KEY_ID")
+type ConfigBasedAWSCredentialsProvider struct {
+	config aws.Config
+}
+
+func (r *ConfigBasedAWSCredentialsProvider) GetUnderlyingProvider() aws.CredentialsProvider {
+	return r.config.Credentials
+}
+
+func (r *ConfigBasedAWSCredentialsProvider) GetRegion() string {
+	return r.config.Region
+}
+
+// Retrieve should be called as late as possible in order to have credentials with latest expiry
+func (r *ConfigBasedAWSCredentialsProvider) Retrieve(ctx context.Context) (AWSCredentials, error) {
+	retrieved, err := r.config.Credentials.Retrieve(ctx)
+	if err != nil {
+		return AWSCredentials{}, err
 	}
+	return AWSCredentials{
+		AWS:         retrieved,
+		EndpointUrl: r.config.BaseEndpoint,
+	}, nil
+}
+
+func NewConfigBasedAWSCredentialsProvider(config aws.Config) AWSCredentialsProvider {
+	return &ConfigBasedAWSCredentialsProvider{config: config}
+}
+
+type StaticAWSCredentialsProvider struct {
+	credentials AWSCredentials
+	region      string
+}
+
+func (s *StaticAWSCredentialsProvider) GetUnderlyingProvider() aws.CredentialsProvider {
+	return credentials.NewStaticCredentialsProvider(s.credentials.AWS.AccessKeyID, s.credentials.AWS.SecretAccessKey,
+		s.credentials.AWS.SessionToken)
+}
+
+func (s *StaticAWSCredentialsProvider) GetRegion() string {
+	return s.region
+}
 
-	awsSecret := creds.SecretAccessKey
-	if awsSecret == "" {
-		awsSecret = os.Getenv("AWS_SECRET_ACCESS_KEY")
+func (s *StaticAWSCredentialsProvider) Retrieve(ctx context.Context) (AWSCredentials, error) {
+	return s.credentials, nil
+}
+
+func NewStaticAWSCredentialsProvider(credentials AWSCredentials, region string) AWSCredentialsProvider {
+	return &StaticAWSCredentialsProvider{
+		credentials: credentials,
+		region:      region,
 	}
+}
 
-	awsRoleArn := creds.AwsRoleArn
-	if awsRoleArn == "" {
-		awsRoleArn = os.Getenv("AWS_ROLE_ARN")
+func getPeerDBAWSEnv(connectorName string, awsKey string) string {
+	return os.Getenv(fmt.Sprintf("PEERDB_%s_AWS_CREDENTIALS_%s", strings.ToUpper(connectorName), awsKey))
+}
```

Comment:
move to `peerdbenv`

Reply:
This is for more fine grain access per connector, the priority of loading would be:
1. Explicit settings from ui/api
2. connector specific aws env
3. peerdb specific aws env (to add via https://github.com/PeerDB-io/peerdb/issues/1641)
4. global aws env